### PR TITLE
Added contribution guidelines

### PR DESCRIPTION
**Pull request by dekarrin**
Added contribution guidelines with link to new Contributor License Agreement. Necessary for legal compliance; does not actually fulfill an issue.

### CODE REVIEW

File: docs/CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+Folks on the discord with an `engineers` or `contributors` role will likely be able to
+quickly help point you in the right direction.
+
+Join the server, let us know that you want to help, and we'll get you something to work
+on.
+
+### GitHub Issues
+
+All of the items in our backlog are located in the GitHub Issues page for the project;
+from there, you should be able to get a pretty good idea of what needs to be worked on.
```
Comment (jghibiki) [inline docs/CONTRIBUTING.md]: maybe insert link to the issues page?

File: docs/CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+It would also be a good idea to notify a project maintainer beforehand if you plan on
+submitting a PR without an associated Issue.
+
+PRs that are not associated with an issue and that do not have reasonable justification given
+in their description are likely to be rejected or sent back for changes.
+
+### Unit Tests and Comments
+
+All PRs that contain changes to code should be accompanied by unit tests that cover the
```
Comment (jghibiki) [inline docs/CONTRIBUTING.md]: Maybe add something about not submitting a pr for an issue you are not assigned to?

File: docs/CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+change. If unit testing does not apply to the situation given in the PR, reasoning should
+be given in the PR description, or a project maintainer should be notified submitting the
```
Comment (jghibiki) [inline docs/CONTRIBUTING.md]: Maybe specify the maintainer should be notified in the PR comments? That way the discord isn't cluttered with individual descriptions of PRs and there is a persistent record of conversations (other than scrolling through the discord history)

File: docs/CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+or class-level docstring explaining what the module and/or class does.
+
+PRs that do not follow the above guidelines are likely to be rejected or sent back for
+changes.
+
+### Style
+
+Please be sure to follow coding style practices established in a project. PRs are likely
+to be rejected or sent back for changes if the style is drastically different from the
+one established in the project.
```
Comment (jghibiki) [inline docs/CONTRIBUTING.md]: Maybe add
> Project maintainers have final say on whether a PR follows the style guidelines

File: docs/CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+If there is a significant reason for the style differing from convention, make sure to
+include the reason either in a code comment or in the PR submission.
+
+For this project, the following guidelines are in effect:
+
+* Use `snake_case` for the names of functions and local variables.
```
Comment (jghibiki) [inline docs/CONTRIBUTING.md]: snake case for module names as well

File: docs/CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+* Use `CamelCase` for the names of classes.
+* In general, PEP-8 should be followed, except for where it does not make sense. The
+explicit cases where this can be ignored:
+  * Global exception Handlers that catch all exceptions are permitted (at the top-level
+  of execution only). Ensure that they properly send the exceptions somewhere (e.g.
+  logging), and that they don't just swallow them.
+  * Shadowing global imports with local identifiers (e.g. it's okay to have a parameter
```
Comment (jghibiki) [inline docs/CONTRIBUTING.md]: maybe add spacing concerns to this list.
e.g.
pep 8 requires that there be two spaces preceding an inline comment. e.g.
`print("hello world")[space]# my comment` - BAD
`print("hello world")[space][space]# my comment` - Good

File: docs/CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+  called `id`).
+
+
+### Architecture Changes
+
+The architecture of the systems in this project should be considered fixed. MSC is not
+opposed to accepting changes to its projects' architectures, but please be aware that
+this will usually not be a simple quick change.
```
Comment (jghibiki) [inline docs/CONTRIBUTING.md]: wording?
> simple or quick change

File: docs/CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+To ensure that all engineers are on the same page, major architectural changes should
+be brought up directly with the engineering team. The easiest way to do that is on
+[the circle's discord server](https://discord.gg/URAA8SF) Discord server.
+
+PRs that necessitate large changes to architecture that are submitted without prior
+discussion will be rejected.
```
Comment (jghibiki) [inline docs/CONTRIBUTING.md]: Maybe add something like:

> The maintaining team may request large changes to be broken up into multiple smaller PRs if the original change set will have overarching effects.

File: docs/CONTRIBUTING.md (hunk 2 above)
Comment (dekarrin) [reply docs/CONTRIBUTING.md]: I do not want to lock that in to policy.

File: docs/CONTRIBUTING.md (hunk 1 above)
Comment (dekarrin) [reply docs/CONTRIBUTING.md]: Ah, good idea!

File: docs/CONTRIBUTING.md (hunk 3 above)
Comment (dekarrin) [reply docs/CONTRIBUTING.md]: That isn't necessarily a particular thing I'd like to set in stone, although we will certainly be prioritizing GH as the go-to location for tracking info, I don't want to discourage conversation from happening in the Discord server.

File: docs/CONTRIBUTING.md (hunk 4 above)
Comment (dekarrin) [reply docs/CONTRIBUTING.md]: I feel like that might be a little redundant, but I have no problem adding it.

File: docs/CONTRIBUTING.md (hunk 5 above)
Comment (dekarrin) [reply docs/CONTRIBUTING.md]: Do you think we should ask for a bias towards no spaces at all in module names?

File: docs/CONTRIBUTING.md (hunk 6 above)
Comment (dekarrin) [reply docs/CONTRIBUTING.md]: Are you saying that we disagree with PEP-8 here? Or that we explicitly agree?

File: docs/CONTRIBUTING.md (hunk 7 above)
Comment (dekarrin) [reply docs/CONTRIBUTING.md]:
```suggestion
this will usually not be a trivial change.
```
?

File: docs/CONTRIBUTING.md (hunk 8 above)
Comment (dekarrin) [reply docs/CONTRIBUTING.md]:
```suggestion
discussion will be rejected; the maintaining team may request further action be taken to split the task into smaller sub-tasks.
```

Ehhhh dunno how I feel about outlining the specific practice... what about this, though?

File: docs/CONTRIBUTING.md (hunk 3 above)
Comment (jghibiki) [reply docs/CONTRIBUTING.md]: yeah, all I meant, is that if you don't update unit tests make a note of why you think you don't need to update the unit tests for the maintainers.

File: docs/CONTRIBUTING.md (hunk 5 above)
Comment (jghibiki) [reply docs/CONTRIBUTING.md]: module names cannot have spaces in them otherwise python cannot import them.

File: docs/CONTRIBUTING.md (hunk 6 above)
Comment (jghibiki) [reply docs/CONTRIBUTING.md]: Just that small nuances like that are a little too tedious to care about. Specifically spacing nuances like that.

Comment (dekarrin) [reply docs/CONTRIBUTING.md]: At least, the comment one. Some other spacing concerns help to establish separation between blocks imo; e.g. the PEP-8 rule on having two new lines between module-level function definitions, so I don't want to rule out all cases of spacing.

File: docs/CONTRIBUTING.md (hunk 5 above)
Comment (jghibiki) [reply docs/CONTRIBUTING.md]: It is more common to see my_engine. I think it is part of pep-8

File: docs/CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+maintainers when possible, but if a PR is specifically connected to an extant issue,
+it is very likely that it will be accepted.
+
+Do note that we use the assignment field of Issues to show who is working on something,
+but the status of that field may not always be up to date. The most recent information
+on what tasks are already in-flight can always be found by asking the active engineers
+at [the circle's discord server](https://discord.gg/URAA8SF).
+
+## Contribution Guidelines
+
```
[the remark this answers is not in the document]
Comment (dekarrin) [reply docs/CONTRIBUTING.md]: Is there something outside of existing Issue templates that we want to cover?

[the remark this answers is not in the document]
Comment (jghibiki) [reply docs/CONTRIBUTING.md]: We do not, at this point, have an Issues template, only a PR template.

[the remark this answers is not in the document]
Comment (dekarrin) [reply docs/CONTRIBUTING.md]: Yeah, let's just do that instead of including it in the contributing guide.

[the remark this answers is not in the document]
Comment (jghibiki) [reply docs/CONTRIBUTING.md]: You will need to do that, as i cannot edit repo settings.

[the remark this answers is not in the document]
Comment (dekarrin) [reply docs/CONTRIBUTING.md]: That's annoying. I'll look into doing that, then.

I'll look into updating permissions so the rest of the team can also do it.

**Pull request merged**